### PR TITLE
[Quant] Fix padded NVFP4 MSE calibration

### LINKED ISSUE #1552: Static NVFP4 weight MSE calibration fails for padded block dimensions
**Before submitting an issue, please make sure it hasn't been already addressed by searching through the [existing and past issues](https://github.com/NVIDIA/Model-Optimizer/issues?q=is%3Aissue).**

## Describe the bug

`mtq.NVFP4_W4A4_WEIGHT_MSE_FP8_SWEEP_CFG` / `nvfp4_mse` fails during static NVFP4 weight MSE calibration when the quantized weight dimension is not divisible by the NVFP4 block size (`16`).

This appears to be a padding/layout mismatch in the MSE calibration path. For a weight with shape `512x60`, static block quantization pads the last dimension to `64`, producing a blocked calibration tensor of shape `2048x16`. The quantized/dequantized result is sliced back to the logical shape `512x60`, then reshaped to `1920x16`, so `F.mse_loss` compares `2048x16` against `1920x16`.

Stack trace ends with:

```text
UserWarning: Using a target size (torch.Size([1920, 16])) that is different to the input size (torch.Size([2048, 16])).
RuntimeError: The size of tensor a (2048) must match the size of tensor b (1920) at non-singleton dimension 0
```

I did not find an existing public issue for this exact padded static NVFP4 MSE shape mismatch.

### Steps/Code to reproduce bug

```python
import copy

import torch
import torch.nn as nn

import modelopt
import modelopt.torch.quantization as mtq


class SingleLinear(nn.Module):
    def __init__(self):
        super().__init__()
        # Last weight dimension is 60, which is not divisible by NVFP4 block size 16.
        self.proj = nn.Linear(60, 512, bias=False)

    def forward(self, x):
        return self.proj(x)


def main():
    print(f"torch: {torch.__version__}")
    print(f"modelopt: {getattr(modelopt, '__version__', '<unknown>')}")

    model = SingleLinear().eval().to(device="cuda", dtype=torch.bfloat16)
    x = torch.randn(2, 60, device="cuda", dtype=torch.bfloat16)

    quant_cfg = copy.deepcopy(mtq.NVFP4_W4A4_WEIGHT_MSE_FP8_SWEEP_CFG)
    mtq.quantize(model, quant_cfg, forward_loop=lambda quantized_model: quantized_model(x))


if __name__ == "__main__":
    main()
```

Observed with ModelOpt `0.44.0`:

```text
torch: 2.8.0+cu128
modelopt: 0.44.0
Inserted 3 quantizers
MSE weight calibration:   0%|          | 0/1 [00:00<?, ?it/s]
UserWarning: Using a target size (torch.Size([1920, 16])) that is different to the input size (torch.Size([2048, 16])).
RuntimeError: The size of tensor a (2048) must match the size of tensor b (1920) at non-singleton dimension 0
```

Sanity checks:

```text
NVFP4_DEFAULT_CFG / Linear(60,512): PASS
NVFP4_W4A4_WEIGHT_MSE_FP8_SWEEP_CFG / Linear(60,512): FAIL
NVFP4_W4A4_WEIGHT_MSE_FP8_SWEEP_CFG / Linear(64,512): PASS
```

I also tested current `main` at commit `b49f9b9e2d747af992d78a3aa7f10efe5a8847e1`. The default Triton fast path passes, but forcing the reference path with:

```bash
MODELOPT_NVFP4_TRITON_SWEEP=0
```

still reproduces the same `2048x16` vs `1920x16` mismatch.

### Expected behavior

Static NVFP4 weight MSE calibration should handle weights whose block dimension is not divisible by `16`, since the quantizer already tracks padding/slicing metadata.

The MSE comparison should use matching layouts, e.g. re-pad the QDQ result before reshaping back to the blocked calibration shape:

```text
logical QDQ: 512x60
re-pad:      512x64
block view:  2048x16
```

At minimum, if this case is intentionally unsupported, ModelOpt should fail early with a clear diagnostic instead of reaching a tensor shape mismatch inside `F.mse_loss`.

### Who can help?

- ?

## System information

- Container used (if applicable): Not using a container; local Python/uv virtual environment
- OS (e.g., Ubuntu 22.04, CentOS 7, Windows 10): Ubuntu 24.04.4 LTS
- CPU architecture (x86_64, aarch64): x86_64
- GPU name (e.g. H100, A100, L40S): NVIDIA GeForce RTX 5080
- GPU memory size: 16303 MiB
- Number of GPUs: 1
- Library versions (if applicable):
  - Python: 3.12.3
  - ModelOpt version or commit hash: 0.44.0; also tested current main at `b49f9b9e2d747af992d78a3aa7f10efe5a8847e1`
  - CUDA: PyTorch CUDA 12.8; NVIDIA driver reports CUDA 13.0
  - PyTorch: 2.8.0+cu128
  - Transformers: 4.57.1
  - TensorRT-LLM: not installed
  - ONNXRuntime: not installed
  - TensorRT: 10.12.0.36
- Any other details that may help:
  - NVIDIA driver: 580.126.20
  - The issue appears specific to static NVFP4 weight MSE / FP8 scale sweep with padded block dimensions.
  - Regular `NVFP4_DEFAULT_CFG` uses dynamic block quantization and does not reproduce this issue on the same `Linear(60,512)` repro.

### PR DESCRIPTION
**Pull request by edmundshieh**
### What does this PR do?

Type of change: Bug fix

Fixes static NVFP4 MSE calibration when the weight's last dimension needs block padding. The reference MSE path now returns fake-quantized tensors to the padded static block layout through the quantizer helper before comparing losses, so padded weights such as `512x60` compare as `2048x16` against `2048x16` instead of `2048x16` against `1920x16`.

The local-Hessian calibration path now reuses the same helper, and a CUDA regression test covers the forced reference sweep for a padded `nn.Linear(60, 512)` weight.

Fixes #1552.

### Usage

```python
import copy
import torch
from torch import nn
import modelopt.torch.quantization as mtq

model = nn.Linear(60, 512, bias=False).eval().cuda().bfloat16()
inputs = torch.randn(2, 60, device="cuda", dtype=torch.bfloat16)
mtq.quantize(
    model,
    copy.deepcopy(mtq.NVFP4_W4A4_WEIGHT_MSE_FP8_SWEEP_CFG),
    forward_loop=lambda model: model(inputs),
)
```

### Testing

- `git diff --check`
- `MODELOPT_NVFP4_TRITON_SWEEP=0 uv run --no-sync python - <<'PY' ...` original padded `Linear(60, 512)` repro: passed
- `uv run --no-sync python -m pytest tests/gpu/torch/quantization/test_nvfp4_static_quantizer_cuda.py`: 11 passed
- `uv run --no-sync python -m pytest tests/unit/torch/quantization/test_mse_calibrator.py`: 14 passed
- `uv run --no-sync pre-commit run --files modelopt/torch/quantization/model_calib.py tests/gpu/torch/quantization/test_nvfp4_static_quantizer_cuda.py`: passed

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: N/A
- Did you write any new necessary tests?: ✅
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: N/A
- Did you get Claude approval on this PR?: N/A

### Additional Information

Related issue: #1552

## Summary by CodeRabbit

* **Refactor**
  * Strengthened quantization calibration robustness with improved state management and explicit restoration mechanisms
  * Unified quantization logic across different calibration methods to eliminate code duplication and improve consistency
  * Refined block-quantization postprocessing behavior for better accuracy

* **Tests**
  * Added comprehensive test coverage validating quantization calibration behavior with padded tensor dimensions and edge cases

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/Model-Optimizer/pull/1557?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by cjluo-nv — APPROVED** [review]
> _Bot review — DM the bot to share feedback._

Targeted bug fix for NVFP4 static MSE calibration when the weight's last dim isn't a multiple of the block size.

- The old `xq.reshape(quantizer._block_reshape_size)` just reshaped by element count and silently dropped the static-block padding (e.g. a `512×60` weight became `1920×16` instead of the padded `2048×16` MseCalibrator was comparing against). Replacing it with `quantizer._process_for_blockquant(xq)` re-applies `_padding` and reshapes to `_original_shape`, matching the calibration block layout. This is the same helper the forward path uses, so layouts can no longer diverge.
- Wrapping the body in `try/finally` correctly restores `_amax` even if quantization throws — a small but worthwhile robustness improvement.
- The local-Hessian path now delegates to `_mse_quant_func`, removing the duplicated copy.
- Regression test exercises the reference (non-Triton) sweep on `Linear(60, 512)` with `MODELOPT_NVFP4_TRITON_SWEEP=0` and asserts the per-block amax has the expected padded count (`512 * 4`).

Small, focused, well-tested, no licensing or design concerns.

File: modelopt/torch/quantization/model_calib.py [hunk 1]
```diff
@@ -395,21 +395,23 @@ def _mse_quant_func(x, amax, quantizer):
...
-            x = quantizer._reset_to_original_shape(x)
-        xq = quantizer(x)
-        if hasattr(quantizer, "_block_reshape_size"):
-            xq = xq.reshape(quantizer._block_reshape_size)
-
-    if original_amax is not None:
-        quantizer._amax = original_amax
-    else:
-        delattr(quantizer, "_amax")
+    try:
```
Comment (shengliangxu) [inline modelopt/torch/quantization/model_calib.py:398]: do we want to have this try block? if error thrown, we run the finally but still the error will be thrown. Do we expect code outside of this function to capture error and move forward?
